lvjit/LuaFunctionBuilder.cpp

```diff
@@ -1633,6 +1689,115 @@ bool Lua::FunctionBuilder::do_forprep(TR::BytecodeBuilder* builder, Instruction
+
+   TR::IlBuilder *lessthan = nullptr;
+   builder->IfThen(&lessthan,
+   builder->       LessThan(
+   builder->                Load("n"),
+   builder->                ConstInt32(0)));
+
+   lessthan->Store("n",
+   lessthan->      ConstInt32(0));
+
+   int b = GETARG_B(instruction) - 1;  /* required results */
+   //if (b < 0) {  this is a compile time constant decision
+   if (b < 0) {
+      //b = n;  /* get all var. arguments */
+      builder->Store("b",
+      builder->      Load("n"));
+
+      //Protect(luaD_checkstack(L, n));
+      builder->Call("jit_checkstack", 2,
+      builder->     Load("L"),
+      builder->     Load("n"));
+      jit_Protect(builder);
+
+      //ra = RA(i);  /* previous call may change the stack */
+      builder->Store("ra", jit_R(builder, GETARG_A(instruction)));
+      //L->top = ra + n;
+      builder->StoreIndirect("lua_State", "top",
+      builder->              Load("L"),
+	  builder->              IndexAt(typeDictionary()->PointerTo(luaTypes.TValue),
+      builder->                      Load("base"),
+      builder->                      Add(
+      builder->                          ConstInt32(GETARG_A(instruction)),
+      builder->                          Load("n"))));
+   } else {
+      builder->Store("b",
+      builder->      ConstInt32(b));
+   }
+
+   //for (j = 0; j < b && j < n; j++)
+   TR::IlBuilder *setvals = nullptr;
+   TR::IlBuilder *loopbreak = nullptr;
+   builder->ForLoopWithBreak(true, "j", &setvals, &loopbreak,
+   builder->                 ConstInt32(0),
+   builder->                 Load("b"),
+   builder->                 ConstInt32(1));
+
+   // j < n break condition
+   TR::IlValue *breakcondition = setvals->LessThan(
+                                 setvals->         Load("j"),
+                                 setvals->         Load("n"));
+   setvals->IfCmpEqualZero(&loopbreak, breakcondition);
+
+   //setobjs2s(L, ra + j, base - n + j);
+   setvals->Store("dest",
+   setvals->      IndexAt(typeDictionary()->PointerTo(luaTypes.TValue),
+   setvals->              Load("base"),
+   setvals->              Add(
+   setvals->                  ConstInt32(GETARG_A(instruction)),
+   setvals->                  Load("j"))));
+   setvals->Store("src",
+   setvals->      IndexAt(typeDictionary()->PointerTo(luaTypes.TValue),
+   setvals->              Load("base"),
+   setvals->              Sub(
+   setvals->                  ConstInt32(0),
+   setvals->                  Sub(
+   setvals->                      Load("n"),
+   setvals->                      Load("j")))));
+   setvals->StoreIndirect("TValue", "value_",
+   setvals->             Load("dest"),
+   setvals->             LoadIndirect("TValue", "value_",
+   setvals->                          Load("src")));
+   setvals->StoreIndirect("TValue", "tt_",
+   setvals->             Load("dest"),
+   setvals->             LoadIndirect("TValue", "tt_",
+   setvals->                          Load("src")));
```

Comment:
This last bit could be simplified with `jit_setobj`. You will have to change the signature from `jit_setobj(TR::BytecodeBuilder* builder, TR::IlValue* dest, TR::IlValue* src)` to `jit_setobj(TR::IlBuilder* builder, TR::IlValue* dest, TR::IlValue* src)` though.